Persist profiles and Firefoxen through reboots

freshstrangemusic opened the pull request.

brennie committed:
Fix features in libfxrecord for tokio

Adding these missing features allows `cargo test -p libfxrecord` to run
successfully.

brennie committed:
Allow multiline docs on message members

The

brennie committed:
Fix cargo clippy lints

brennie committed:
Add RequestManager for creating and resuming requests

brennie committed:
Use Wait instead of bool in integration tests

The previous changes that replaced the `bool` arugment in
`RecorderProto::SendResumeRequest` with a `Wait` enum did not update the
integration tests.

brennie committed:
Propagate errors in profile creation to recorder

brennie committed:
Use the RequestManager in the RunnerProto

brennie committed:
Update all integration tests to use a RequestManager

brennie committed:
Clean up failed requests

If a request fails before the runner restarts, we now clean up any
downloaded files (Firefoxen, profiles, etc).

brennie committed:
Do not use a temporary directory for request state

brennie committed:
Clean up all requests after serving resume requests

Requests are not intended to be served out of order. fxrunner must only
send a single new request followed by a corresponding resume request. We
now clean up the request directory after each resume request.

squarewave commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,176 @@
+}
+/// A trait for creating and validating requests.
+#[async_trait]
+pub trait RequestManager {
```
I think "request" may be a bit vague or even misleading for all of this. Maybe something like "trial" or "session" or "transaction". This doesn't need to change now, but it's definitely confusing coming in.

keeferrourke replied:
I agree; especially the idea of "cleaning up a request" later in this file is a bit confusing.

squarewave commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,176 @@
+        let mut path = self.path.clone();
+
+        for _ in 0..TRIES {
+            for _ in 0..8 {
```
I think this would read cleaner, and have a nearly identical result, if you did something like:

```
let bytes = thread_rng().gen::<[u8; 6]>();
request_id = base64::encode(bytes);
```

I'm not sure if there's a reason why you use such a low number of bytes, though. I would be inclined to use more bytes and just fail if the path already exists, since you could make it astronomically unlikely.

squarewave commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,176 @@
+
+#[async_trait]
+impl RequestManager for FsRequestManager {
+    async fn new_request(&self) -> Result<RequestInfo<'static>, NewRequestError> {
```
I think the implementation of `new_request` needs just a little bit of commenting explaining why we're doing what we're doing.

squarewave commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,176 @@
+        &self,
+        request_id: &'a str,
+    ) -> Result<RequestInfo<'a>, ResumeRequestError> {
+        if request_id.chars().any(|c| !c.is_ascii_alphanumeric()) {
```
(Of course you'd need to remove this bit if you switched to base64 and just check if it failed to decode)

squarewave commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,176 @@
+            });
+        }
+
+        if !self.path.join("profile").is_dir_async().await {
```
It's not clear to me why we do this as async and the above as sync.

squarewave commented on libfxrecord/src/net/message.rs:
```diff
@@ -20,10 +20,22 @@
 //!        BarMsg {
+//!            /// The contents of the message.
+//!            ///
+//!            /// The type of this field is `i32`.
```
Nit: `u32`

squarewave reviewed (approved):
Seems reasonable overall - just have a few comments in line.

brennie committed:
Cleanup completed resumed requests

The runner now will cleanup resumed requests as they are completed
(either successfully or unsucessfully). Additionally, the
`FsRequestManager` will cleanup any requests that it cannot resume.

brennie committed:
Handle requests that produce errors without exiting

The runner no longer exits when encountering an error after handling a
request that produced an error. This allows us to recover from, e.g.,
invalid requests, which are not fatal and should not cause the runner to
exit.

brennie commented:
I added some patches to this stack to handle things I ran into when I was addressing this feedback.

keeferrourke commented on fxrunner/src/bin/main.rs:
```diff
@@ -70,22 +83,35 @@ async fn fxrunner(log: Logger, options: Options, config: Config) -> Result<(), B
+
+            // We aren't restarting, which means we handled a resume request. We
+            // only expect a signle pending request at a time, so the request
+            // directory *should* be empty. If it, then isn't empty it.
```
Nit: grammar. I think the last part of this comment should read "If it isn't, then empty it"?

keeferrourke commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,212 @@
+        request_id: &'a str,
+    ) -> Result<RequestInfo<'a>, ResumeRequestError> {
+        if request_id.len() != REQUEST_ID_LEN
+            || request_id.chars().any(|c| !c.is_ascii_alphanumeric())
```
nit: it might help local readability if this logic is extracted to a helper. Something like`fn validate_request_id(id: str) -> bool`?

keeferrourke commented on fxrunner/src/lib/fs.rs:
```diff
@@ -0,0 +1,33 @@
+use async_trait::async_trait;
+use tokio::fs::metadata;
+
+/// An extension trait for Path to add async versions of its metadata helpers.
+#[async_trait]
+pub trait PathExt {
+    /// An async version of `std::path::Path::is_file`.
+    async fn is_file_async(&self) -> bool;
+
+    /// An async version of `std::path::Path::is_dir`.
+    async fn is_dir_async(&self) -> bool;
+}
```
This is nice :D

keeferrourke commented on fxrunner/src/lib/proto.rs:
```diff
@@ -264,7 +316,10 @@ where
-        let unzip_path = self.working_dir.join("profile");
+        // It is possible that the profile contains a top-level directory, in
+        // which case we don't want to directly extract to
+        // `request_info.path.join("profile")`.
```
What do we want to do instead?

keeferrourke commented on fxrunner/src/lib/request.rs:
```diff
@@ -0,0 +1,212 @@
+        Ok(request_info)
+    }
+
+    async fn ensure_profile_dir<'a>(
```
nit: maybe `ensure_profile_dir_exists`?

keeferrourke reviewed (approved):
LGTM, just a few nits :)

squarewave commented on fxrunner/src/bin/main.rs:
```diff
@@ -70,22 +83,35 @@ async fn fxrunner(log: Logger, options: Options, config: Config) -> Result<(), B
             info!(log, "Client disconnected");
+
+            // We aren't restarting, which means we handled a resume request. We
+            // only expect a signle pending request at a time, so the request
```
Nit: s/signle/single

squarewave reviewed (approved):
LGTM
